datasetComparison/src/main/scala/za/co/absa/hermes/datasetComparison/DatasetComparator.scala

```diff
@@ -265,11 +265,12 @@ class DatasetComparator(dataFrameReference: DataFrame,
    * @return Returns a DataFrame with key column appended
    */
   private def addKeyColumn(selector: List[Column], df: DataFrame, cmpUniqueColumn: String): DataFrame = {
-    if (keys.nonEmpty) {
-      df.withColumn(cmpUniqueColumn, md5(concat_ws("|", keys.map(col).toSeq: _*)))
+    val columnsToConcatenate = if (keys.nonEmpty) {
+      keys.map(col(_).cast("string")).toSeq
     } else {
-      df.withColumn(cmpUniqueColumn, md5(concat_ws("|", selector: _*)))
+      selector.map(_.cast("string"))
     }
+    df.withColumn(cmpUniqueColumn, md5(concat_ws("|", columnsToConcatenate: _*)))
```

Comment:
You said that you will use `concat` instead of `concat_ws`? Can you add some tests for a few corner cases such as nulls, etc?

Reply:
Yeah, I left an old description, will update it. As you can see "nothing changed" I just added a cast to string. 
And I don't have private functions covered in unit tests anyway. But other tests cover the likes of nulls.

Reply:
Added a struct test